docs(tproxy): restructure and expand transparent proxy documentation (#2064)
kumahq/kuma-website: app/_src/networking/transparent-proxy/kubernetes.md

### Configuration in ConfigMap (experimental)

{% warning %}
Because this feature impacts multiple underlying components, it is considered experimental. Use it with caution. {% if site.mesh_product_name == "Kuma" %} If you encounter any unexpected behavior or issues, please [**contact us**](/community) and [**submit an issue on GitHub**](https://github.com/kumahq/kuma/issues/new/choose). Your feedback is essential in helping us improve this feature. {% endif %}
{% endwarning %}

Until {{ Kuma }} 2.9, transparent proxy settings could only be modified through the [control plane runtime configuration](#control-plane-runtime-configuration) and [annotations](#annotations), which had several limitations:

- Not all settings were available through both methods; some could only be adjusted with annotations, while others were limited to the runtime configuration.
- Control plane runtime settings were applied globally to all workloads with injected data planes, making it hard to customize settings for specific groups of workloads.

> - Annotation-based settings had to be added to each workload individually, making it difficult to manage the same configuration across multiple workloads.

Author: turned this around to advantages

Here is an example of how to modify parts of this configuration during installation:

{% cpinstall transparent-proxy-config-configmap-auto-creation-examples %}
{{ transparentProxy }}.configMap.config.enabled=true
{{ transparentProxy }}.configMap.config.redirect.outbound.excludePortsForIPs=10\.1\.0\.254\,172\.10\.1.254
{{ transparentProxy }}.configMap.config.redirect.verbose=false
{% endcpinstall %}

{% tip %}
{{ Kuma }} uses a single configuration structure for transparent proxy settings across all components. For the full configuration schema, see the [Helm values.yaml reference]({{ docs }}/reference/kuma-cp/#helm-valuesyaml), particularly under the `{{ transparentProxy }}.configMap.config` path. More details on each setting are available in the [Transparent Proxy Configuration reference]({{ docs }}/reference/transparent-proxy-configuration/#full-reference).
{% endtip %}

> #### Custom ConfigMap name

Author: addressed

The name of the ConfigMap is defined by the `{{ transparentProxy }}.configMap.name` setting. By default, this name is set to `kuma-transparent-proxy-config`.

To apply a custom name for all workloads, you can modify the `{{ transparentProxy }}.configMap.name` setting during installation:

{% cpinstall transparent-proxy-config-configmap-custom-name-globally %}
{{ transparentProxy }}.configMap.name=custom-name
{% endcpinstall %}

> To use a different resource for specific workloads, apply the `traffic.kuma.io/transparent-proxy-configmap-name` annotation to those workloads. For example:

Author: actually changed this section so this should be resolved

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: demo-app
  annotations:
    traffic.kuma.io/transparent-proxy-configmap-name: custom-name
...
```

#### ConfigMap lookup order

{{ Kuma }} looks for the ConfigMap as follows:

1. **If annotation is set**: Search the workload's namespace for the ConfigMap specified by `traffic.kuma.io/transparent-proxy-configmap-name`.

2. **If annotation is set but ConfigMap not found**: Search the `{{ kuma-system }}` namespace for the same ConfigMap.

3. **If no annotation or previous steps failed**: Look for the ConfigMap with the default name in the workload's namespace.

> 4. **If still not found**: Search the `{{ kuma-system }}` namespace for the ConfigMap with the default name.

Author: Addressed via https://github.com/kumahq/kuma/pull/13012 and the change in the mentioned point

{% endtip %}

#### Settings restricted to control plane runtime configuration

Some transparent proxy settings **can only be changed through the control plane’s runtime configuration** because they are shared with other {{ Kuma }} components. These settings handle essential tasks like creating [`Dataplane` resources]({{ docs }}/production/dp-config/dpp/) for workloads and setting up the [`kuma-dp` sidecar container]({{ docs }}/production/dp-config/dpp-on-kubernetes/) alongside them. For example, the DNS port used to redirect traffic is shared between the transparent proxy and the `kuma-dp run` command in the sidecar container. Keeping these settings consistent across workloads prevents misconfigurations. These settings are:

- `runtime.kubernetes.injector.sidecarContainer.redirectPortInbound`
- `runtime.kubernetes.injector.sidecarContainer.redirectPortOutbound`
- `runtime.kubernetes.injector.sidecarContainer.uid`
- `runtime.kubernetes.injector.builtinDNS.enabled`
- `runtime.kubernetes.injector.builtinDNS.port`

{% danger %}
⚠️ IMPORTANT
{:.custom-block-title}

> If you change these settings, but with values different from the runtime configuration, they will be ignored, and a warning will appear in the control plane logs (except for DNS-related settings).

Reviewer: ```suggestion
If you change these settings by Annotation or ConfigMap the change will be ignored, and a warning will appear in the control plane logs (except for DNS-related settings).
```

Author: addressed

{% assign kuma-system = site.mesh_namespace %}
{% assign controlPlane = site.set_flag_values_prefix | append: "controlPlane" %}
{% assign transparentProxy = site.set_flag_values_prefix | append: "transparentProxy" %}
{% assign kuma-control-plane = site.mesh_cp_name %}
{% assign kuma-control-plane-config = kuma-control-plane | append: "-config" %}

In Kubernetes mode, transparent proxy is automatically set up through the `kuma-init` container or [Kuma CNI]({{ docs }}/production/dp-config/cni/). By default, it intercepts all incoming and outgoing traffic and routes it through the `kuma-dp` sidecar container, so no changes to the application code are needed.

{{ Kuma }} works smoothly with [Kubernetes DNS for Services and Pods](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/) and provides its own [Kuma DNS]({{ docs }}/networking/dns/), which is especially helpful in multi-zone setups for cross-zone service discovery.

In this mode, {{ Kuma }} requires the transparent proxy to be enabled, so it **cannot be turned off**.

## Configuration

The default configuration works well for most scenarios, but there are cases where adjustments are needed.

> In Kubernetes mode, {{ Kuma }} there are three methods to adjust the configuration. Each can be used on its own or combined with others if needed.

Reviewer: ```suggestion
In Kubernetes mode, {{ Kuma }} has three methods to adjust the configuration. Each can be used on its own or combined with others if needed.
```

Author: addressed

{% warning %}
It’s best to stick to one method whenever possible. Using more than one can make things more complicated and harder to troubleshoot, as it may not be clear where each setting comes from. If you need to combine methods, check the [**Order of Precedence** section](#order-of-precedence) to see what the final configuration will look like based on the priority of each setting.
{% endwarning %}

<!-- vale Google.Headings = NO -->
### Control Plane Runtime Configuration
<!-- vale Google.Headings = YES -->

The control plane runtime configuration is a section of {{ Kuma }}'s configuration that contains important settings for how the control plane operates, including options for the transparent proxy and other key components.

Some transparent proxy settings can be adjusted here, and **for certain settings, this is the only place they can be modified**.

> Currently, it’s best to use the control plane runtime configuration as the main place to store transparent proxy settings that apply to all [workloads](https://kubernetes.io/docs/concepts/workloads/). In the future, once the [Configuration in ConfigMap](#configuration-in-configmap-experimental) feature is fully supported, we’ll recommend using ConfigMaps for these settings. However, a few restricted settings that are rarely customized will still need to be in the control plane’s runtime configuration.

Reviewer: ```suggestion
Currently, it’s best to use the control plane runtime configuration as the main place to store transparent proxy settings that apply to all [workloads](https://kubernetes.io/docs/concepts/workloads/).
```

I think we should only be talking about current state of project in docs. This should be a snapshot of knowledge around specific Kuma version. This might happen, but we could also switch our approach and remove config maps

Author: addressed

{% endtip %}

#### Settings restricted to control plane runtime configuration

Some transparent proxy settings **can only be changed through the control plane’s runtime configuration** because they are shared with other {{ Kuma }} components. These settings handle essential tasks like creating [`Dataplane` resources]({{ docs }}/production/dp-config/dpp/) for workloads and setting up the [`kuma-dp` sidecar container]({{ docs }}/production/dp-config/dpp-on-kubernetes/) alongside them. For example, the DNS port used to redirect traffic is shared between the transparent proxy and the `kuma-dp run` command in the sidecar container. Keeping these settings consistent across workloads prevents misconfigurations. These settings are:

- `runtime.kubernetes.injector.sidecarContainer.redirectPortInbound`
- `runtime.kubernetes.injector.sidecarContainer.redirectPortOutbound`
- `runtime.kubernetes.injector.sidecarContainer.uid`
- `runtime.kubernetes.injector.builtinDNS.enabled`
- `runtime.kubernetes.injector.builtinDNS.port`

{% danger %}
⚠️ IMPORTANT
{:.custom-block-title}

> If you change these settings, but with values different from the runtime configuration, they will be ignored, and a warning will appear in the control plane logs (except for DNS-related settings).

Reviewer: If you change these settings by Annotation or ConfigMap the change will be ignored, and a warning will appear in the control plane logs (except for DNS-related settings).

If `builtinDNS.enabled` or `builtinDNS.port` is updated using deprecated annotations while [Configuration in ConfigMap](#configuration-in-configmap-experimental) is disabled, the changes may still apply, potentially causing DNS redirection issues. This could prevent `kuma-dp` from starting the DNS server or listening on the correct port, leading to environment disruptions.
{% enddanger %}

#### Modifying control plane runtime configuration

For instructions on modifying the control plane configuration, see the [Modifying the Configuration]({{ docs }}/documentation/configuration/#modifying-the-configuration) section in the [Control Plane Configuration]({{ docs }}/documentation/configuration/) documentation.

> #### Overriding list-based configuration settings

Reviewer: I think that this whole subsection should be moved to https://deploy-preview-2064--kuma.netlify.app/docs/2.9.x/documentation/configuration/#modifying-the-configuration it is not transparent proxy specific

Author: addressed

Until {{ Kuma }} 2.9, transparent proxy settings could only be modified through the [Control Plane Runtime Configuration](#control-plane-runtime-configuration) and [Annotations](#annotations), which had several limitations:

- Not all settings were available through both methods; some could only be adjusted with annotations, while others were limited to the runtime configuration.
- Control plane runtime settings were applied globally to all workloads with injected data planes, making it hard to customize settings for specific groups of workloads.
- Annotation-based settings had to be added to each workload individually, making it difficult to manage the same configuration across multiple workloads.

To address these issues, {{ Kuma }} now supports storing transparent proxy settings in dedicated ConfigMaps. These ConfigMaps can be applied at the namespace or individual workload level, offering greater flexibility and easier configuration.

To enable this feature, set `{{ transparentProxy }}.configMap.enabled` during installation:

{% cpinstall transparent-proxy-config-configmap-enabling %}
{{ transparentProxy }}.configMap.enabled=true
{% endcpinstall %}

{% warning %}

> If you set `{{ transparentProxy }}.configMap.config` to an empty value, it will override `{{ transparentProxy }}.configMap.enabled` and disable the feature, even if `{{ transparentProxy }}.configMap.enabled` is set to `true`.

Reviewer: this should just enable configMaps

Author: It was implemented like that, but with https://github.com/kumahq/kuma/pull/11988 it will just enable configMaps as you suggested

Author: wrapped this with `if_version` to not show it in 2.10.x+

Here is an example of how to modify parts of this configuration during installation:

{% cpinstall transparent-proxy-config-configmap-auto-creation-examples %}
{{ transparentProxy }}.configMap.config.enabled=true
{{ transparentProxy }}.configMap.config.redirect.outbound.excludePortsForIPs=10\.1\.0\.254\,172\.10\.1.254
{{ transparentProxy }}.configMap.config.redirect.verbose=false
{% endcpinstall %}

You can apply the same method of providing these values as explained in the **Store the desired settings in the `{{ kuma-control-plane-config }}` ConfigMap** option in the [**Modifying Control Plane Runtime Configuration**](#modifying-control-plane-runtime-configuration) section.

{% tip %}
For more information about the structure of this configuration, see the [Kuma Control Plane Helm Values]({{ docs }}/reference/kuma-cp/#helm-valuesyaml) documentation, particularly under the `{{ transparentProxy }}.configMap.config` path. You can also refer to the [Transparent Proxy Configuration Reference]({{ docs }}/reference/transparent-proxy-configuration/) for additional details.
{% endtip %}

<!-- vale Google.Headings = NO -->

> #### ConfigMap lookup strategy

Reviewer: this section could be after instruction on how to change name

Author: addressed

<!-- vale Google.Headings = YES -->

{{ Kuma }} follows a specific order when searching for the ConfigMap:

> - If the workload is annotated with `traffic.kuma.io/transparent-proxy-configmap-name`, {{ Kuma }} first looks for the resource with the specified name in the workload’s namespace. If not found, it will then check the `{{ kuma-system }}` namespace for the same ConfigMap.

Reviewer: this could be splitted to separate points to make order straightforward to read

Author: addressed

- If the workload is not annotated, or the ConfigMap specified in the annotation is not found, {{ Kuma }} will search for the resource named in the `{{ transparentProxy }}.configMap.name` setting. It first looks in the workload's namespace, and if not found, it checks the `{{ kuma-system }}` namespace.

{% warning %}

> The ConfigMap in the `{{ kuma-system }}` namespace is required for proper operation, so it must be present even if custom ones are used in all individual workload namespaces.

Reviewer: why we need default configmap?

```sh
kubectl annotate pods <pod_name> "traffic.kuma.io/transparent-proxy-configmap-name=custom-name"
```

This allows flexibility, enabling different workloads to use distinct ConfigMap settings while maintaining a global default.

### Annotations

Kubernetes annotations can be applied to individual workloads to modify the transparent proxy configuration. These annotations allow fine-tuning of specific behaviors for a single workload without affecting others.

{% tip %}
Before {{ Kuma }} 2.9, annotations were the only way to modify the transparent proxy configuration on a per-workload basis. In {{ Kuma }} 2.9, the [Configuration in ConfigMap](#configuration-in-configmap-experimental) feature was introduced, offering a more efficient way to configure groups of workloads. When this feature is no longer experimental, it will become the recommended approach for managing transparent proxy settings across multiple workloads. Annotations will still be useful for fine-tuning specific configurations when you need an individual workload to behave differently from the others.
{% endtip %}

> Below you can find a list of annotations that can be used to configure specific transparent proxy settings. 

Reviewer: this list should be moved to reference section

Author: addressed
